rust/bridge/node/futures/src/future.rs

```diff
@@ -0,0 +1,205 @@
+//
+// Copyright 2020-2021 Signal Messenger, LLC.
+// SPDX-License-Identifier: AGPL-3.0-only
+//
+
+use neon::prelude::*;
+use std::future::Future;
+use std::marker::PhantomData;
+use std::mem;
+use std::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
+use std::pin::Pin;
+use std::sync::{Arc, Mutex, Weak};
+use std::task::{Poll, Waker};
+
+use crate::result::*;
+
+mod builder;
+pub use builder::JsFutureBuilder;
+
+/// A transformation to convert a scoped JavaScript result (fulfillment or rejection) to an unscoped Rust result.
+trait JsFutureCallback<T> =
+    for<'a> FnOnce(&mut FunctionContext<'a>, JsPromiseResult<'a>) -> T + 'static + Send;
+
+/// The possible states of a [JsFuture].
+enum JsFutureState<T> {
+    /// The future is waiting for resolution.
+    Waiting {
+        transform: Box<dyn JsFutureCallback<T>>,
+        waker: Option<Waker>,
+    },
+    /// The future has been resolved (and transformed).
+    ///
+    /// If there was a panic during that transform, it will be caught and resumed at the `await` point.
+    Complete(std::thread::Result<T>),
+    /// The result has been returned from `poll`.
```

Comment:
The error state doesn't actually correspond to rejection; it corresponds to a panic during the transform, which we forward to the `await`ing context. (Note that it's a `thread::Result` rather than a JsPromiseResult.) It's possible to split that up ahead of time but `catch_unwind` hands us a `thread::Result` anyway, so it's simpler not to bother.